In [1]:
import numpy as np
import pandas as pd

In [2]:
loan_df=pd.read_csv(r"C:\Users\Prasanna\Documents\IIT Hyderabad\Sem 1\ML\Assignment_3\loan_train.csv")

In [3]:
loan_df

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0       573354     737474      10000        10000       9950.00000   
1       476321     603324      15000        15000      14800.00000   
2       451484     556265       2000         2000       2000.00000   
3      1018129    1246557      35000        35000      33951.84413   
4       800018    1005270      14000        14000      14000.00000   
...        ...        ...        ...          ...              ...   
24994   450579     554340       5500         5500       5500.00000   
24995   788033     991661      11450        11450      11450.00000   
24996   568459     731299      24000        24000      21100.31424   
24997   397594     428786      14000        14000      11411.12089   
24998   660367     844601      13400        13400      13400.00000   

             term int_rate  installment grade sub_grade  ...  \
0       36 months    7.51%       311.11     A        A4  ...   
1       36 months    8.94%       476.58     A        A5  ...   
2       36 months   13.57%        67.94     C        C3  ...   
3       60 months   20.89%       944.71     F        F1  ...   
4       60 months   17.49%       351.64     D        D5  ...   
...           ...      ...          ...   ...       ...  ...   
24994   36 months   14.96%       190.55     D        D2  ...   
24995   36 months    8.49%       361.40     A        A5  ...   
24996   60 months   20.53%       642.96     G        G2  ...   
24997   36 months   17.58%       503.19     F        F2  ...   
24998   36 months   13.06%       451.89     C        C2  ...   

      num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75  \
0                    NaN                NaN            NaN               NaN   
1                    NaN                NaN            NaN               NaN   
2                    NaN                NaN            NaN               NaN   
3                    NaN                NaN            NaN               NaN   
4                    NaN                NaN            NaN               NaN   
...                  ...                ...            ...               ...   
24994                NaN                NaN            NaN               NaN   
24995                NaN                NaN            NaN               NaN   
24996                NaN                NaN            NaN               NaN   
24997                NaN                NaN            NaN               NaN   
24998                NaN                NaN            NaN               NaN   

      pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort  \
0                      0.0       0.0             NaN               NaN   
1                      0.0       0.0             NaN               NaN   
2                      0.0       0.0             NaN               NaN   
3                      2.0       0.0             NaN               NaN   
4                      0.0       0.0             NaN               NaN   
...                    ...       ...             ...               ...   
24994                  0.0       0.0             NaN               NaN   
24995                  0.0       0.0             NaN               NaN   
24996                  0.0       0.0             NaN               NaN   
24997                  1.0       0.0             NaN               NaN   
24998                  0.0       0.0             NaN               NaN   

      total_bc_limit total_il_high_credit_limit  
0                NaN                        NaN  
1                NaN                        NaN  
2                NaN                        NaN  
3                NaN                        NaN  
4                NaN                        NaN  
...              ...                        ...  
24994            NaN                        NaN  
24995            NaN                        NaN  
24996            NaN                        NaN  
24997            NaN                        NaN  
24

In [4]:
#removing all the data points with loan status as "Current" as it is not relevant to this problem
loan_df.drop(loan_df.index[loan_df['loan_status'] == 'Current'], inplace=True)

In [5]:
#dropping all the columns with all the values as Null (i.e NaN)
loan_df.dropna(how='all', axis=1, inplace=True)

In [6]:
loan_df

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0      573354     737474      10000        10000       9950.00000   36 months   
1      476321     603324      15000        15000      14800.00000   36 months   
2      451484     556265       2000         2000       2000.00000   36 months   
4      800018    1005270      14000        14000      14000.00000   60 months   
5      471391     595223      12000        12000      11900.00000   36 months   
...       ...        ...        ...          ...              ...         ...   
24994  450579     554340       5500         5500       5500.00000   36 months   
24995  788033     991661      11450        11450      11450.00000   36 months   
24996  568459     731299      24000        24000      21100.31424   60 months   
24997  397594     428786      14000        14000      11411.12089   36 months   
24998  660367     844601      13400        13400      13400.00000   36 months   

      int_rate  installment grade sub_grade  ... last_pymnt_amnt  \
0        7.51%       311.11     A        A4  ...          311.11   
1        8.94%       476.58     A        A5  ...        15114.03   
2       13.57%        67.94     C        C3  ...          101.78   
4       17.49%       351.64     D        D5  ...         3943.27   
5        8.94%       381.26     A        A5  ...        11204.30   
...        ...          ...   ...       ...  ...             ...   
24994   14.96%       190.55     D        D2  ...         1122.20   
24995    8.49%       361.40     A        A5  ...          382.90   
24996   20.53%       642.96     G        G2  ...        16500.39   
24997   17.58%       503.19     F        F2  ...         3857.88   
24998   13.06%       451.89     C        C2  ...          475.29   

      last_credit_pull_d collections_12_mths_ex_med  policy_code  \
0                 11-Jul                        0.0            1   
1                 10-Feb                        0.0            1   
2                 11-Dec                        0.0            1   
4                 16-May                        0.0            1   
5                 10-May                        0.0            1   
...                  ...                        ...          ...   
24994             12-Jun                        0.0            1   
24995             16-May                        0.0            1   
24996             16-May                        0.0            1   
24997             15-Jun                        0.0            1   
24998             14-Feb                        0.0            1   

      application_type acc_now_delinq chargeoff_within_12_mths delinq_amnt  \
0           INDIVIDUAL              0                      0.0           0   
1           INDIVIDUAL              0                      0.0           0   
2           INDIVIDUAL              0                      0.0           0   
4           INDIVIDUAL              0                      0.0           0   
5           INDIVIDUAL              0                      0.0           0   
...                ...            ...                      ...         ...   
24994       INDIVIDUAL              0                      0.0           0   
24995       INDIVIDUAL              0                      0.0           0   
24996       INDIVIDUAL              0                      0.0           0   
24997       INDIVIDUAL              0                      0.0           0   
24998       INDIVIDUAL              0                      0.0           0   

      pub_rec_bankruptcies tax_liens  
0                      0.0       0.0  
1                      0.0       0.0  
2                      0.0       0.0  
4                      0.0       0.0  
5                      0.0       0.0  
...                    ...       ...  
24994                  0.0       0.0  
24995                  0.0       0.0  
24996                  0.0       0.0  
24997                  1.0       0.0  
24998                  0.0       0.0  

Hence we could see the total number of columns has reduced to half and also some of the rows were removed with loan status as "Current" (around 900 rows)

In [7]:
#information about all the columns can be obtained by
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 56 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          24301 non-null  int64  
 1   member_id                   24301 non-null  int64  
 2   loan_amnt                   24301 non-null  int64  
 3   funded_amnt                 24301 non-null  int64  
 4   funded_amnt_inv             24301 non-null  float64
 5   term                        24301 non-null  object 
 6   int_rate                    24301 non-null  object 
 7   installment                 24301 non-null  float64
 8   grade                       24301 non-null  object 
 9   sub_grade                   24301 non-null  object 
 10  emp_title                   24295 non-null  object 
 11  emp_length                  24261 non-null  object 
 12  home_ownership              24301 non-null  object 
 13  annual_inc                  243

### Removing the unnecessary columns by manually by eyeballing

In [8]:
#reasons for removing these columns are given in a separate file.
loan_df.drop(['id','member_id','emp_title'], axis=1, inplace=True)
loan_df.drop(['issue_d','pymnt_plan','url','desc','title','zip_code'], axis=1, inplace=True)
loan_df.drop(['earliest_cr_line','mths_since_last_delinq','mths_since_last_record','initial_list_status','out_prncp','out_prncp_inv'], axis=1, inplace=True)
loan_df.drop(['last_pymnt_d','last_credit_pull_d','collections_12_mths_ex_med','policy_code','application_type','acc_now_delinq','chargeoff_within_12_mths','delinq_amnt','tax_liens'], axis=1, inplace=True)

In [9]:
loan_df.drop('sub_grade', axis=1, inplace=True)

In [10]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                24301 non-null  int64  
 1   funded_amnt              24301 non-null  int64  
 2   funded_amnt_inv          24301 non-null  float64
 3   term                     24301 non-null  object 
 4   int_rate                 24301 non-null  object 
 5   installment              24301 non-null  float64
 6   grade                    24301 non-null  object 
 7   emp_length               24261 non-null  object 
 8   home_ownership           24301 non-null  object 
 9   annual_inc               24301 non-null  float64
 10  verification_status      24301 non-null  object 
 11  loan_status              24301 non-null  object 
 12  purpose                  24301 non-null  object 
 13  addr_state               24301 non-null  object 
 14  dti                   

Hence, we have reduced the total number of columns to 31. But this cannot be directly used to train the model as there are missing values and also some columns containing categorical variables.

The columns which contain missing values are:
1. emp_length
2. revol_util
3. pub_rec_bankruptcies

### Dealing with missing values

In [11]:
loan_df['pub_rec_bankruptcies'].unique()

array([ 0.,  1., nan,  2.])

In [12]:
#fillling NaN values of pub_rec_bankruptcies column with forward fill
loan_df['pub_rec_bankruptcies'] = loan_df['pub_rec_bankruptcies'].ffill()

In [13]:
loan_df['emp_length'].unique()

array(['10+ years', '1 year', '4 years', '5 years', '7 years', '2 years',
       '8 years', '< 1 year', '3 years', '9 years', '6 years', nan],
      dtype=object)

In [14]:
#fillling NaN values of emp_length column with forward fill
loan_df['emp_length'] = loan_df['emp_length'].ffill()

The column 'emp_length' contains numbers as well as strings such as '+' ,'<', 'years'. Let us remove them to extract only the numbers 

In [15]:
loan_df['emp_length'] = loan_df['emp_length'].map(lambda x: x.rstrip('+ years'))
loan_df['emp_length'] = loan_df['emp_length'].replace(['< 1'],'0.5')
loan_df['emp_length'].astype(str).astype(float)
loan_df['emp_length'].unique()

array(['10', '1', '4', '5', '7', '2', '8', '0.5', '3', '9', '6'],
      dtype=object)

In [16]:
loan_df['emp_length'] = loan_df['emp_length'].astype(float)
loan_df['emp_length'].unique()

array([10. ,  1. ,  4. ,  5. ,  7. ,  2. ,  8. ,  0.5,  3. ,  9. ,  6. ])

In [17]:
loan_df['revol_util']

0         1.20%
1           14%
2        34.40%
4         9.40%
5         2.80%
          ...  
24994    58.20%
24995       34%
24996    77.20%
24997       95%
24998    41.80%
Name: revol_util, Length: 24301, dtype: object

In [18]:
#fillling NaN values of revol_util column with forward fill
loan_df['revol_util'] = loan_df['revol_util'].ffill()
loan_df['revol_util'] = loan_df['revol_util'].map(lambda x: x.rstrip('%'))

In [19]:
loan_df['revol_util'] = loan_df['revol_util'].astype(float)
loan_df['revol_util'].unique()

array([ 1.2 , 14.  , 34.4 , ..., 62.31,  4.85, 13.56])

In [20]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                24301 non-null  int64  
 1   funded_amnt              24301 non-null  int64  
 2   funded_amnt_inv          24301 non-null  float64
 3   term                     24301 non-null  object 
 4   int_rate                 24301 non-null  object 
 5   installment              24301 non-null  float64
 6   grade                    24301 non-null  object 
 7   emp_length               24301 non-null  float64
 8   home_ownership           24301 non-null  object 
 9   annual_inc               24301 non-null  float64
 10  verification_status      24301 non-null  object 
 11  loan_status              24301 non-null  object 
 12  purpose                  24301 non-null  object 
 13  addr_state               24301 non-null  object 
 14  dti                   

In [21]:
#checking if there are missing values in the entire data set
loan_df.isnull().values.sum()

0

Hence there are no null values in the entire data set. Now we have to convert categorical features to numerical values

In [22]:
loan_df

loan_amnt  funded_amnt  funded_amnt_inv        term int_rate  \
0          10000        10000       9950.00000   36 months    7.51%   
1          15000        15000      14800.00000   36 months    8.94%   
2           2000         2000       2000.00000   36 months   13.57%   
4          14000        14000      14000.00000   60 months   17.49%   
5          12000        12000      11900.00000   36 months    8.94%   
...          ...          ...              ...         ...      ...   
24994       5500         5500       5500.00000   36 months   14.96%   
24995      11450        11450      11450.00000   36 months    8.49%   
24996      24000        24000      21100.31424   60 months   20.53%   
24997      14000        14000      11411.12089   36 months   17.58%   
24998      13400        13400      13400.00000   36 months   13.06%   

       installment grade  emp_length home_ownership  annual_inc  ...  \
0           311.11     A        10.0           RENT     30000.0  ...   
1           476.58     A         1.0           RENT    147000.0  ...   
2            67.94     C         4.0            OWN     36000.0  ...   
4           351.64     D         5.0       MORTGAGE     50000.0  ...   
5           381.26     A         7.0       MORTGAGE    294000.0  ...   
...            ...   ...         ...            ...         ...  ...   
24994       190.55     D         7.0       MORTGAGE     52000.0  ...   
24995       361.40     A         7.0           RENT     40000.0  ...   
24996       642.96     G        10.0       MORTGAGE     74454.0  ...   
24997       503.19     F        10.0       MORTGAGE     86000.0  ...   
24998       451.89     C         0.5           RENT     62040.0  ...   

      total_acc   total_pymnt total_pymnt_inv total_rec_prncp  total_rec_int  \
0            48   2247.210000         2235.98         1509.91         350.87   
1            17  15112.760000        14911.26        15000.00         112.76   
2            10   2354.966827         2354.97         2000.00         354.97   
4            27  20804.230020        20804.23        14000.00        6804.23   
5            21  12344.811770        12241.94        12000.00         344.81   
...         ...           ...             ...             ...            ...   
24994        37   6825.066783         6825.07         5500.00        1325.07   
24995        15  13010.206030        13010.21        11450.00        1560.21   
24996        24  33765.068310        26978.96        24000.00        9765.07   
24997        26  17917.012900        14191.06        14000.00        3917.01   
24998        14  16268.440550        16268.44        13400.00        2868.44   

       total_rec_late_fee  recoveries  collection_recovery_fee  \
0                     0.0      386.43                     3.96   
1                     0.0        0.00                     0.00   
2                     0.0        0.00                     0.00   
4                     0.0        0.00                     0.00   
5                     0.0        0.00                     0.00   
...                   ...         ...                      ...   
24994                 0.0        0.00                     0.00   
24995                 0.0        0.00                     0.00   
24996                 0.0        0.00                     0.00   
24997                 0.0        0.00                     0.00   
24998                 0.0        0.00                     0.00   

       last_pymnt_amnt  pub_rec_bankruptcies  
0               311.11                   0.0  
1             15114.03                   0.0  
2               101.78                   0.0  
4              3943.27                   0.0  
5             11204.30                   0.0  
...                ...                   ...  
24994          1122.20                   0.0  
24995           382.90                   0.0  
24996         16500.39                   0.0  
24997          3857.88                   1.0  
24998           47

Here, we could see that the column 'int_rate' has % symbol which is unnecessary. Let us remove that.

In [23]:
loan_df['int_rate'] = loan_df['int_rate'].map(lambda x: x.rstrip('%'))
loan_df['int_rate'] = loan_df['int_rate'].astype(float)
loan_df['int_rate']

0         7.51
1         8.94
2        13.57
4        17.49
5         8.94
         ...  
24994    14.96
24995     8.49
24996    20.53
24997    17.58
24998    13.06
Name: int_rate, Length: 24301, dtype: float64

In [24]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                24301 non-null  int64  
 1   funded_amnt              24301 non-null  int64  
 2   funded_amnt_inv          24301 non-null  float64
 3   term                     24301 non-null  object 
 4   int_rate                 24301 non-null  float64
 5   installment              24301 non-null  float64
 6   grade                    24301 non-null  object 
 7   emp_length               24301 non-null  float64
 8   home_ownership           24301 non-null  object 
 9   annual_inc               24301 non-null  float64
 10  verification_status      24301 non-null  object 
 11  loan_status              24301 non-null  object 
 12  purpose                  24301 non-null  object 
 13  addr_state               24301 non-null  object 
 14  dti                   

### Converting categorical to numerical

The columns that contain the categorical attributes include:
3. term
6. grade
8. home_ownership
10. verification_status
11. loan_status (label column)
12. purpose
13. addr_state

In [25]:
loan_df['grade'].unique()

array(['A', 'C', 'D', 'B', 'E', 'F', 'G'], dtype=object)

In [26]:
#converting the categorical to numerical attributes
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
loan_df['grade']=le.fit_transform(loan_df.grade)

In [27]:
loan_df['grade'].unique()

array([0, 2, 3, 1, 4, 5, 6])

In [28]:
loan_df['term'].unique()

array([' 36 months', ' 60 months'], dtype=object)

In [29]:
loan_df['term']=le.fit_transform(loan_df.term)
loan_df['term'].unique()

array([0, 1])

In [30]:
loan_df['home_ownership'].unique()

array(['RENT', 'OWN', 'MORTGAGE', 'OTHER'], dtype=object)

In [31]:
loan_df['home_ownership']=le.fit_transform(loan_df.home_ownership)
loan_df['home_ownership'].unique()

array([3, 2, 0, 1])

In [32]:
loan_df['verification_status'].unique()

array(['Source Verified', 'Not Verified', 'Verified'], dtype=object)

In [33]:
loan_df['verification_status']=le.fit_transform(loan_df.verification_status)
loan_df['verification_status'].unique()

array([1, 0, 2])

In [34]:
loan_df['purpose'].unique()

array(['home_improvement', 'other', 'major_purchase',
       'debt_consolidation', 'credit_card', 'small_business', 'wedding',
       'medical', 'moving', 'car', 'educational', 'vacation', 'house',
       'renewable_energy'], dtype=object)

In [35]:
loan_df['purpose']=le.fit_transform(loan_df.purpose)
loan_df['purpose'].unique()

array([ 4,  9,  6,  2,  1, 11, 13,  7,  8,  0,  3, 12,  5, 10])

In [36]:
loan_df['addr_state'].unique()

array(['NJ', 'GA', 'FL', 'CA', 'MO', 'MI', 'OR', 'CT', 'PA', 'IL', 'MN',
       'AZ', 'NY', 'NC', 'WI', 'UT', 'VA', 'MD', 'CO', 'TX', 'OK', 'OH',
       'AL', 'NV', 'AK', 'DC', 'VT', 'SC', 'WA', 'DE', 'KS', 'HI', 'MA',
       'KY', 'WY', 'NH', 'NM', 'LA', 'WV', 'AR', 'RI', 'IA', 'TN', 'MT',
       'NE', 'SD', 'MS', 'ME', 'ID', 'IN'], dtype=object)

In [37]:
loan_df['addr_state']=le.fit_transform(loan_df.addr_state)
loan_df['addr_state'].unique()

array([30, 10,  9,  4, 24, 22, 36,  6, 37, 14, 23,  3, 33, 27, 47, 43, 44,
       20,  5, 42, 35, 34,  1, 32,  0,  7, 45, 39, 46,  8, 16, 11, 19, 17,
       49, 29, 31, 18, 48,  2, 38, 12, 41, 26, 28, 40, 25, 21, 13, 15])

Converting the label column 'loan_status' as  numerical and then placing it as last column is given below:

In [38]:
loan_df['loan_status'].unique()

array(['Charged Off', 'Fully Paid'], dtype=object)

In [39]:
loan_df['loan_status']=le.fit_transform(loan_df.loan_status)
loan_df['loan_status'].unique()

array([0, 1])

In [40]:
#moving the loan_status to last column
col_name="loan_status"
label = loan_df.pop(col_name)
loan_df.insert(30, col_name, label)

In [41]:
loan_df.head()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  \
0      10000        10000           9950.0     0      7.51       311.11   
1      15000        15000          14800.0     0      8.94       476.58   
2       2000         2000           2000.0     0     13.57        67.94   
4      14000        14000          14000.0     1     17.49       351.64   
5      12000        12000          11900.0     0      8.94       381.26   

   grade  emp_length  home_ownership  annual_inc  ...   total_pymnt  \
0      0        10.0               3     30000.0  ...   2247.210000   
1      0         1.0               3    147000.0  ...  15112.760000   
2      2         4.0               2     36000.0  ...   2354.966827   
4      3         5.0               0     50000.0  ...  20804.230020   
5      0         7.0               0    294000.0  ...  12344.811770   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          2235.98          1509.91         350.87                 0.0   
1         14911.26         15000.00         112.76                 0.0   
2          2354.97          2000.00         354.97                 0.0   
4         20804.23         14000.00        6804.23                 0.0   
5         12241.94         12000.00         344.81                 0.0   

   recoveries  collection_recovery_fee  last_pymnt_amnt  pub_rec_bankruptcies  \
0      386.43                     3.96           311.11                   0.0   
1        0.00                     0.00         15114.03                   0.0   
2        0.00                     0.00           101.78                   0.0   
4        0.00                     0.00          3943.27                   0.0   
5        0.00                     0.00         11204.30                   0.0   

   loan_status  
0            0  
1            1  
2            1  
4            1  
5            1  

[5 rows x 31 columns]

In [42]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                24301 non-null  int64  
 1   funded_amnt              24301 non-null  int64  
 2   funded_amnt_inv          24301 non-null  float64
 3   term                     24301 non-null  int32  
 4   int_rate                 24301 non-null  float64
 5   installment              24301 non-null  float64
 6   grade                    24301 non-null  int32  
 7   emp_length               24301 non-null  float64
 8   home_ownership           24301 non-null  int32  
 9   annual_inc               24301 non-null  float64
 10  verification_status      24301 non-null  int32  
 11  purpose                  24301 non-null  int32  
 12  addr_state               24301 non-null  int32  
 13  dti                      24301 non-null  float64
 14  delinq_2yrs           

In [43]:
loan_df.isnull().values.sum()

0

**Hence, our train data set is now ready!!**

### Preparing Test data set in the similar manner above:

In [44]:
test_df=pd.read_csv(r"C:\Users\Prasanna\Documents\IIT Hyderabad\Sem 1\ML\Assignment_3\loan_test.csv")

In [45]:
test_df

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1001205    1227186       4200         4200           4200.0   
1       850106    1061893       1325         1325           1325.0   
2       444243     541365      10000        10000           9975.0   
3       741007     938805       6000         6000           6000.0   
4       572468     736352      25000        25000          25000.0   
...        ...        ...        ...          ...              ...   
14713    98276      98268       5400         5400            200.0   
14714    93277      93254       3000         3000            950.0   
14715    92676      92671       5000         5000            150.0   
14716    90395      90390       5000         5000           1325.0   
14717    90376      89243       5000         5000            650.0   

             term int_rate  installment grade sub_grade  ...  \
0       36 months    6.62%       128.96     A        A2  ...   
1       36 months   13.49%        44.96     C        C2  ...   
2       36 months    8.94%       317.72     A        A5  ...   
3       60 months   16.49%       147.48     D        D3  ...   
4       36 months   15.21%       869.21     D        D2  ...   
...           ...      ...          ...   ...       ...  ...   
14713   36 months    7.75%       168.60     A        A3  ...   
14714   36 months    8.70%        94.98     B        B1  ...   
14715   36 months    8.07%       156.84     A        A4  ...   
14716   36 months    8.07%       156.84     A        A4  ...   
14717   36 months    7.43%       155.38     A        A2  ...   

      num_tl_90g_dpd_24m num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75  \
0                    NaN                NaN            NaN               NaN   
1                    NaN                NaN            NaN               NaN   
2                    NaN                NaN            NaN               NaN   
3                    NaN                NaN            NaN               NaN   
4                    NaN                NaN            NaN               NaN   
...                  ...                ...            ...               ...   
14713                NaN                NaN            NaN               NaN   
14714                NaN                NaN            NaN               NaN   
14715                NaN                NaN            NaN               NaN   
14716                NaN                NaN            NaN               NaN   
14717                NaN                NaN            NaN               NaN   

      pub_rec_bankruptcies tax_liens tot_hi_cred_lim total_bal_ex_mort  \
0                      0.0       0.0             NaN               NaN   
1                      0.0       0.0             NaN               NaN   
2                      0.0       0.0             NaN               NaN   
3                      0.0       0.0             NaN               NaN   
4                      0.0       0.0             NaN               NaN   
...                    ...       ...             ...               ...   
14713                  NaN       NaN             NaN               NaN   
14714                  NaN       NaN             NaN               NaN   
14715                  NaN       NaN             NaN               NaN   
14716                  NaN       NaN             NaN               NaN   
14717                  NaN       NaN             NaN               NaN   

      total_bc_limit total_il_high_credit_limit  
0                NaN                        NaN  
1                NaN                        NaN  
2                NaN                        NaN  
3                NaN                        NaN  
4                NaN                        NaN  
...              ...                        ...  
14713            NaN                        NaN  
14714            NaN                        NaN  
14715            NaN                        NaN  
14716            NaN                        NaN  
14

In [46]:
#removing all the data points with loan status as "Current" as it is not relevant to this problem
test_df.drop(test_df.index[test_df['loan_status'] == 'Current'], inplace=True)

In [47]:
#dropping all the columns with all the values as Null (i.e NaN)
test_df.dropna(how='all', axis=1, inplace=True)

In [48]:
test_df

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv  \
0      1001205    1227186       4200         4200           4200.0   
1       850106    1061893       1325         1325           1325.0   
2       444243     541365      10000        10000           9975.0   
3       741007     938805       6000         6000           6000.0   
4       572468     736352      25000        25000          25000.0   
...        ...        ...        ...          ...              ...   
14713    98276      98268       5400         5400            200.0   
14714    93277      93254       3000         3000            950.0   
14715    92676      92671       5000         5000            150.0   
14716    90395      90390       5000         5000           1325.0   
14717    90376      89243       5000         5000            650.0   

             term int_rate  installment grade sub_grade  ... last_pymnt_amnt  \
0       36 months    6.62%       128.96     A        A2  ...          132.82   
1       36 months   13.49%        44.96     C        C2  ...           44.96   
2       36 months    8.94%       317.72     A        A5  ...         9337.10   
3       60 months   16.49%       147.48     D        D3  ...          147.02   
4       36 months   15.21%       869.21     D        D2  ...          960.31   
...           ...      ...          ...   ...       ...  ...             ...   
14713   36 months    7.75%       168.60     A        A3  ...          169.09   
14714   36 months    8.70%        94.98     B        B1  ...           97.45   
14715   36 months    8.07%       156.84     A        A4  ...          157.31   
14716   36 months    8.07%       156.84     A        A4  ...            0.00   
14717   36 months    7.43%       155.38     A        A2  ...            0.00   

      last_credit_pull_d collections_12_mths_ex_med  policy_code  \
0                 16-May                        0.0            1   
1                 14-Sep                        0.0            1   
2                 16-May                        0.0            1   
3                 16-May                        0.0            1   
4                 16-May                        0.0            1   
...                  ...                        ...          ...   
14713             13-Dec                        NaN            1   
14714             10-Jul                        NaN            1   
14715              7-Jun                        NaN            1   
14716              7-Jun                        NaN            1   
14717              7-Jun                        NaN            1   

      application_type acc_now_delinq chargeoff_within_12_mths delinq_amnt  \
0           INDIVIDUAL              0                      0.0           0   
1           INDIVIDUAL              0                      0.0           0   
2           INDIVIDUAL              0                      0.0           0   
3           INDIVIDUAL              0                      0.0           0   
4           INDIVIDUAL              0                      0.0           0   
...                ...            ...                      ...         ...   
14713       INDIVIDUAL              0                      NaN           0   
14714       INDIVIDUAL              0                      NaN           0   
14715       INDIVIDUAL              0                      NaN           0   
14716       INDIVIDUAL              0                      NaN           0   
14717       INDIVIDUAL              0                      NaN           0   

      pub_rec_bankruptcies tax_liens  
0                      0.0       0.0  
1                      0.0       0.0  
2                      0.0       0.0  
3                      0.0       0.0  
4                      0.0       0.0  
...                    ...       ...  
14713                  NaN       NaN  
14714                  NaN       NaN  
14715                  NaN       NaN  
14716                  NaN       NaN  
14717                  NaN 

In [49]:
#information about all the columns can be obtained by
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14276 entries, 0 to 14717
Data columns (total 56 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   id                          14276 non-null  int64  
 1   member_id                   14276 non-null  int64  
 2   loan_amnt                   14276 non-null  int64  
 3   funded_amnt                 14276 non-null  int64  
 4   funded_amnt_inv             14276 non-null  float64
 5   term                        14276 non-null  object 
 6   int_rate                    14276 non-null  object 
 7   installment                 14276 non-null  float64
 8   grade                       14276 non-null  object 
 9   sub_grade                   14276 non-null  object 
 10  emp_title                   11896 non-null  object 
 11  emp_length                  13283 non-null  object 
 12  home_ownership              14276 non-null  object 
 13  annual_inc                  142

In [50]:
#reasons for removing these columns are given in a separate file.
test_df.drop(['id','member_id','emp_title'], axis=1, inplace=True)
test_df.drop(['issue_d','pymnt_plan','url','desc','title','zip_code'], axis=1, inplace=True)
test_df.drop(['earliest_cr_line','mths_since_last_delinq','mths_since_last_record','initial_list_status','out_prncp','out_prncp_inv'], axis=1, inplace=True)
test_df.drop(['last_pymnt_d','last_credit_pull_d','collections_12_mths_ex_med','policy_code','application_type','acc_now_delinq','chargeoff_within_12_mths','delinq_amnt','tax_liens'], axis=1, inplace=True)

In [51]:
test_df.drop('sub_grade', axis=1, inplace=True)

In [52]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14276 entries, 0 to 14717
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                14276 non-null  int64  
 1   funded_amnt              14276 non-null  int64  
 2   funded_amnt_inv          14276 non-null  float64
 3   term                     14276 non-null  object 
 4   int_rate                 14276 non-null  object 
 5   installment              14276 non-null  float64
 6   grade                    14276 non-null  object 
 7   emp_length               13283 non-null  object 
 8   home_ownership           14276 non-null  object 
 9   annual_inc               14276 non-null  float64
 10  verification_status      14276 non-null  object 
 11  loan_status              14276 non-null  object 
 12  purpose                  14276 non-null  object 
 13  addr_state               14276 non-null  object 
 14  dti                   

In [53]:
#fillling NaN values of pub_rec_bankruptcies column with forward fill
test_df['pub_rec_bankruptcies'] = test_df['pub_rec_bankruptcies'].ffill()

#fillling NaN values of emp_length column with forward fill
test_df['emp_length'] = test_df['emp_length'].ffill()

In [54]:
#Further processing with 'emp_length' column
test_df['emp_length'] = test_df['emp_length'].map(lambda x: x.rstrip('+ years'))
test_df['emp_length'] = test_df['emp_length'].replace(['< 1'],'0.5')
test_df['emp_length'] = test_df['emp_length'].astype(float)
test_df['emp_length'].unique()

array([ 6. ,  0.5,  4. , 10. ,  5. ,  8. ,  3. ,  1. ,  7. ,  9. ,  2. ])

In [55]:
#fillling NaN values of revol_util column with forward fill
test_df['revol_util'] = test_df['revol_util'].ffill()
test_df['revol_util'] = test_df['revol_util'].map(lambda x: x.rstrip('%'))
test_df['revol_util'] = test_df['revol_util'].astype(float)
test_df['revol_util'].unique()

array([11.6 , 11.9 , 46.4 , ..., 56.26, 29.77,  6.75])

In [56]:
test_df['int_rate'] = test_df['int_rate'].map(lambda x: x.rstrip('%'))
test_df['int_rate'] = test_df['int_rate'].astype(float)
test_df['int_rate']

0         6.62
1        13.49
2         8.94
3        16.49
4        15.21
         ...  
14713     7.75
14714     8.70
14715     8.07
14716     8.07
14717     7.43
Name: int_rate, Length: 14276, dtype: float64

In [57]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14276 entries, 0 to 14717
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                14276 non-null  int64  
 1   funded_amnt              14276 non-null  int64  
 2   funded_amnt_inv          14276 non-null  float64
 3   term                     14276 non-null  object 
 4   int_rate                 14276 non-null  float64
 5   installment              14276 non-null  float64
 6   grade                    14276 non-null  object 
 7   emp_length               14276 non-null  float64
 8   home_ownership           14276 non-null  object 
 9   annual_inc               14276 non-null  float64
 10  verification_status      14276 non-null  object 
 11  loan_status              14276 non-null  object 
 12  purpose                  14276 non-null  object 
 13  addr_state               14276 non-null  object 
 14  dti                   

In [58]:
#checking if there are missing values in the entire test data
test_df.isnull().values.sum()

0

In [59]:
#converting the categorical to numerical attributes
test_df['grade']=le.fit_transform(test_df.grade)
test_df['term']=le.fit_transform(test_df.term)
test_df['home_ownership']=le.fit_transform(test_df.home_ownership)
test_df['verification_status']=le.fit_transform(test_df.verification_status)
test_df['purpose']=le.fit_transform(test_df.purpose)
test_df['addr_state']=le.fit_transform(test_df.addr_state)

In [60]:
#Processing label column:
test_df['loan_status']=le.fit_transform(test_df.loan_status)
c_name="loan_status"
target = test_df.pop(c_name)
test_df.insert(30, c_name, target)

In [61]:
test_df.head()

loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  \
0       4200         4200           4200.0     0      6.62       128.96   
1       1325         1325           1325.0     0     13.49        44.96   
2      10000        10000           9975.0     0      8.94       317.72   
3       6000         6000           6000.0     1     16.49       147.48   
4      25000        25000          25000.0     0     15.21       869.21   

   grade  emp_length  home_ownership  annual_inc  ...   total_pymnt  \
0      0         6.0               4     69000.0  ...   4642.374867   
1      2         0.5               4     24000.0  ...   1476.640000   
2      0         4.0               0     75000.0  ...  10287.347790   
3      3        10.0               4     60000.0  ...   8848.343924   
4      3         4.0               4    135000.0  ...  31293.587380   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          4642.37          4200.00         442.37             0.00000   
1          1476.64          1145.61         287.60            14.93661   
2         10261.63         10000.00         287.35             0.00000   
3          8848.34          6000.00        2848.34             0.00000   
4         31293.59         25000.00        6293.59             0.00000   

   recoveries  collection_recovery_fee  last_pymnt_amnt  pub_rec_bankruptcies  \
0         0.0                    0.000           132.82                   0.0   
1        28.5                    0.285            44.96                   0.0   
2         0.0                    0.000          9337.10                   0.0   
3         0.0                    0.000           147.02                   0.0   
4         0.0                    0.000           960.31                   0.0   

   loan_status  
0            1  
1            0  
2            1  
3            1  
4            1  

[5 rows x 31 columns]

In [62]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14276 entries, 0 to 14717
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   loan_amnt                14276 non-null  int64  
 1   funded_amnt              14276 non-null  int64  
 2   funded_amnt_inv          14276 non-null  float64
 3   term                     14276 non-null  int32  
 4   int_rate                 14276 non-null  float64
 5   installment              14276 non-null  float64
 6   grade                    14276 non-null  int32  
 7   emp_length               14276 non-null  float64
 8   home_ownership           14276 non-null  int32  
 9   annual_inc               14276 non-null  float64
 10  verification_status      14276 non-null  int32  
 11  purpose                  14276 non-null  int32  
 12  addr_state               14276 non-null  int32  
 13  dti                      14276 non-null  float64
 14  delinq_2yrs           

In [63]:
test_df.isnull().values.sum()

0

**Hence, our test data set is now ready!!**

### Splitting X_train, Y_train, X_test and Y_test

In [64]:
X_train = loan_df.iloc[:, :-1].values
Y_train= loan_df.iloc[:, -1].values.reshape(-1,1)

In [65]:
X_train.shape, Y_train.shape

((24301, 30), (24301, 1))

In [66]:
X_test = test_df.iloc[:, :-1].values
Y_test= test_df.iloc[:, -1].values.reshape(-1,1)

In [67]:
X_test.shape, Y_test.shape

((14276, 30), (14276, 1))

In [83]:
from sklearn.ensemble import GradientBoostingClassifier
model1 = GradientBoostingClassifier(n_estimators=30, learning_rate=1, max_features=5, max_depth=5, random_state=0)
model1.fit(X_train, Y_train.ravel())
Y_pred=model1.predict(X_test)
from sklearn.metrics import precision_score, recall_score
ps=precision_score(Y_test,Y_pred)
rs=recall_score(Y_test,Y_pred)
print("Accuracy score :",model1.score(X_test,Y_test))
print("Precision: ",ps)
print("Recall: ",rs)

Accuracy score : 0.9947464275707482
Precision:  0.9954762296430334
Recall:  0.9983502433391075


### Let us analyse the different accuracies for increasing number of trees:

In [84]:
est_list=[10,20,30,40,50]
for n_est in est_list:
    model2=GradientBoostingClassifier(n_estimators=n_est, learning_rate=1, max_features=5, max_depth=5, random_state=0)
    model2.fit(X_train, Y_train.ravel())
    Y_pred=model2.predict(X_test)
    ps=precision_score(Y_test,Y_pred)
    rs=recall_score(Y_test,Y_pred)
    print("The accuracy for",n_est,"estimators with learning rate=1, max_features=5 and max_depth=5 is",model2.score(X_test,Y_test),"Precision: ",ps,"Recall: ",rs)

The accuracy for 10 estimators with learning rate=1, max_features=5 and max_depth=5 is 0.9932053796581676 Precision:  0.9940026289845547 Recall:  0.998020292006929
The accuracy for 20 estimators with learning rate=1, max_features=5 and max_depth=5 is 0.9936957130848977 Precision:  0.9947372748951566 Recall:  0.9978553163408397
The accuracy for 30 estimators with learning rate=1, max_features=5 and max_depth=5 is 0.9947464275707482 Precision:  0.9954762296430334 Recall:  0.9983502433391075
The accuracy for 40 estimators with learning rate=1, max_features=5 and max_depth=5 is 0.994536284673578 Precision:  0.9950678175092479 Recall:  0.9985152190051967
The accuracy for 50 estimators with learning rate=1, max_features=5 and max_depth=5 is 0.9949565704679182 Precision:  0.9953144266337854 Recall:  0.9987626825043306


#### Therefore the accuracy increases in most of the cases when the number of trees increases

Let us analyse the different accuracies for different learning rate:

In [85]:
#let us fix the n_estimators as 30 as it gave the second highest accuracy
lr_list = [0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1]
for lr in lr_list:
    model3=GradientBoostingClassifier(n_estimators=30, learning_rate=lr, max_features=5, max_depth=5, random_state=0)
    model3.fit(X_train, Y_train.ravel())
    Y_pred=model3.predict(X_test)
    ps=precision_score(Y_test,Y_pred)
    rs=recall_score(Y_test,Y_pred)
    print("The accuracy for 30 estimators with learning rate=",lr,",max_features=5 and max_depth=5 is",model3.score(X_test,Y_test),"Precision: ",ps,"Recall: ",rs)
    

The accuracy for 30 estimators with learning rate= 0.05 ,max_features=5 and max_depth=5 is 0.9789857102829924 Precision:  0.9758512436609514 Recall:  1.0
The accuracy for 30 estimators with learning rate= 0.075 ,max_features=5 and max_depth=5 is 0.9835388063883441 Precision:  0.9809839779899661 Recall:  1.0
The accuracy for 30 estimators with learning rate= 0.1 ,max_features=5 and max_depth=5 is 0.9869010927430653 Precision:  0.9850455136540962 Recall:  0.9997525365008662
The accuracy for 30 estimators with learning rate= 0.25 ,max_features=5 and max_depth=5 is 0.9934855701877277 Precision:  0.9929519750860515 Recall:  0.9994225851686876
The accuracy for 30 estimators with learning rate= 0.5 ,max_features=5 and max_depth=5 is 0.9949565704679182 Precision:  0.9948263118994827 Recall:  0.9992576095025983
The accuracy for 30 estimators with learning rate= 0.75 ,max_features=5 and max_depth=5 is 0.9942560941440179 Precision:  0.9947407346536281 Recall:  0.9985152190051967
The accuracy for 

Therefore, the accuracy increases with increase in learning rate

In [86]:
#let us fix n_estimators=30 and learning_rate=0.5
l=[1,2,3,4,5]
for i in l:
    model4=GradientBoostingClassifier(n_estimators=30, learning_rate=0.5, max_features=i, max_depth=i, random_state=0)
    model4.fit(X_train, Y_train.ravel())
    model4.fit(X_train, Y_train.ravel())
    Y_pred=model4.predict(X_test)
    ps=precision_score(Y_test,Y_pred)
    rs=recall_score(Y_test,Y_pred)
    print("The accuracy for 30 estimators with learning rate= 0.5, max_features=",i, "and max_depth=",i,"is",model4.score(X_test,Y_test),"Precision: ",ps,"Recall: ",rs)

The accuracy for 30 estimators with learning rate= 0.5, max_features= 1 and max_depth= 1 is 0.9501961333706921 Precision:  0.9446002805049089 Recall:  1.0
The accuracy for 30 estimators with learning rate= 0.5, max_features= 2 and max_depth= 2 is 0.9820678061081536 Precision:  0.9797849114579121 Recall:  0.9995050730017322
The accuracy for 30 estimators with learning rate= 0.5, max_features= 3 and max_depth= 3 is 0.991454188848417 Precision:  0.9910794664047794 Recall:  0.9989276581704198
The accuracy for 30 estimators with learning rate= 0.5, max_features= 4 and max_depth= 4 is 0.9939058559820678 Precision:  0.9934404722859954 Recall:  0.9994225851686876
The accuracy for 30 estimators with learning rate= 0.5, max_features= 5 and max_depth= 5 is 0.9949565704679182 Precision:  0.9948263118994827 Recall:  0.9992576095025983


Hence the best test accuracy is obtained when there are 30 estimators, learning rate=0.5, max_features=5 and max_depth=5 which is 0.9949565704679182

### Comparing with simple decision tree

In [87]:
from sklearn.tree import DecisionTreeClassifier
dtree=DecisionTreeClassifier(criterion='entropy',max_features=5, max_depth=5)
dtree.fit(X_train,Y_train)
Y_pred=dtree.predict(X_test)
ps=precision_score(Y_test,Y_pred)
rs=recall_score(Y_test,Y_pred)
acc=dtree.score(X_test,Y_test)
print("Accuracy: ",acc,"Precision: ",ps,"Recall: ",rs)

Accuracy:  0.9618240403474363 Precision:  0.961421967160848 Recall:  0.9949682421842778


Therefore, from the above accuracies, we can see that the Gradient boosting (accuracy:0.9949565704679182) performs much better than the decision tree (accuracy: 0.9618240403474363)